### PR TITLE
AZP: Add multi HCAs iodemo test

### PR DESCRIPTION
**Pull request by avildema**
## What
add test for multi HPCs test for iodemo

### CODE REVIEW

**Review by huaxiangfan — COMMENTED** [review]
1. arp sysctl needs to be configured staticly:
```
sysctl -w net.ipv4.conf.all.arp_filter=0
sysctl -w net.ipv4.conf.default.arp_filter=0
sysctl -w net.ipv4.conf.all.arp_ignore=1
sysctl -w net.ipv4.conf.default.arp_ignore=1
sysctl -w net.ipv4.conf.all.arp_announce=1
sysctl -w net.ipv4.conf.default.arp_announce=1
sysctl -w net.ipv4.conf.all.rp_filter=2
sysctl -w net.ipv4.conf.default.rp_filter=2
sysctl -w net.ipv4.conf.all.arp_accept=1
sysctl -w net.ipv4.conf.default.arp_accept=1
```
2. iodemo_analyzer.py should not skip any error except 'Connection reset by remote peer' for this case

File: buildlib/pr/io_demo/az-stage-io-demo.yaml [hunk 1]
```diff
@@ -72,6 +79,100 @@ steps:
...
+- bash: |
+    set -x
+    source $(workspace)/buildlib/az-helpers.sh
+    IFS=, read host1 host2 <<< $(agent_hosts)
+    set_subnet_mask ${{parameters.roce_iface}} 255.255.0.0
+    ssh $host2 "source ${PWD}/$(workspace)/buildlib/az-helpers.sh && set_subnet_mask ${{parameters.roce_iface}} 255.255.0.0"
+
+  condition: eq(variables['single_hca'], 'No')
+  timeoutInMinutes: 15
+  displayName: Set subnet Mask
```
[the remark this answers is not in the document]
Comment (avildema) [reply buildlib/pr/io_demo/az-stage-io-demo.yaml:93]: thanks, will add

File: buildlib/pr/io_demo/az-stage-io-demo.yaml [hunk 2]
```diff
@@ -72,6 +79,100 @@ steps:
...
+    IFS=, read host1 host2 <<< $(agent_hosts)
+
+    ip1_h1=$(get_ip ${iface1})
+    ip2_h1=$(get_ip ${iface2})
+    ip1_h2=$(ssh $host2 "source ${PWD}/$(workspace)/buildlib/az-helpers.sh && get_ip ${iface1}" )
+    ip2_h2=$(ssh $host2 "source ${PWD}/$(workspace)/buildlib/az-helpers.sh && get_ip ${iface2}" )
+
+    export UCX_TLS=${{ parameters.iodemo_tls }}
+    export UCX_RNDV_THRESH=4k
+    export LD_LIBRARY_PATH=$(workspace)/install/lib:$LD_LIBRARY_PATH
```
[the remark this answers is not in the document]
Comment (avildema) [reply buildlib/pr/io_demo/az-stage-io-demo.yaml:120]: done

File: buildlib/pr/io_demo/az-stage-io-demo.yaml [hunk 3]
```diff
@@ -72,6 +79,100 @@ steps:
...
+      ${PWD}/$(io_demo_exe) \
+        -d 512:524288 -P 2 -i 0 -w 16 -t 60 ${{ parameters.iodemo_args }} ${ip2_h2}:20000 &>> \
+        $(workspace)/${{ parameters.name }}/client_0.log &
+
+    ssh ${host2} " \
+      IODEMO_ROLE=client_1 LD_LIBRARY_PATH=${PWD}/$(workspace)/install/lib:$LD_LIBRARY_PATH UCX_NET_DEVICES=${net1} \
+        stdbuf -e0 -o0 timeout -s 9 ${{ parameters.duration }} \
+        ${PWD}/$(io_demo_exe) \
+          -d 512:524288 -P 2 -w 16 -t 60 ${{ parameters.iodemo_args }} ${ip2_h1}:20001 &>> \
+        ${PWD}/$(workspace)/${{ parameters.name }}/client_1.log &"
```
[the remark this answers is not in the document]
Comment (avildema) [reply buildlib/pr/io_demo/az-stage-io-demo.yaml:155]: What's do you mean?
need to remove other clients?

File: buildlib/pr/io_demo/az-stage-io-demo.yaml [hunk 4]
```diff
@@ -72,14 +81,69 @@ steps:
...
+    sleep 5
+    # clients
+
+    ssh ${host2} " \
+      IODEMO_ROLE=client_0 LD_LIBRARY_PATH=${PWD}/$(workspace)/install/lib:$LD_LIBRARY_PATH UCX_NET_DEVICES=${net1} \
+      UCX_CM_USE_ALL_DEVICES=n UCX_RNDV_THRESH=4k \
+        stdbuf -e0 -o0 timeout -s 9 ${{ parameters.duration }} \
+        ${PWD}/$(io_demo_exe) \
+          -d 512:524288 -i 0 -P 2 -w 16 -t 60 ${{ parameters.iodemo_args }} ${ip1_h1}:20000 ${ip2_h1}:20000 &>> \
+        ${PWD}/$(workspace)/${{ parameters.name }}/client_2.log &"
```
Comment (changchengx) [inline buildlib/pr/io_demo/az-stage-io-demo.yaml:125]: @avildema Will ```iodemo_analyzer``` parse ```server_0.log``` and ```client_2.log``` file to check whether there're unexpected error?

Comment (avildema) [reply buildlib/pr/io_demo/az-stage-io-demo.yaml:125]: will parce all logs

**Comment by avildema** [comment]
@yosefe maybe squash and merge?

File: buildlib/az-network-corrupter.sh [hunk 5]
```diff
@@ -25,10 +27,14 @@ sleep ${initial_delay}
 
 for i in $(seq 1 ${cycles}); do
     echo "#$i Put it down! And sleep ${downtime}"
-    ${manager_script} --one -d ${manager_script_dir}/hosts --host $(hostname) -a off -i ${interface}
+    for interface in ${interfaces} ; do
+        ${manager_script} --one -d ${manager_script_dir}/hosts --host $(hostname) -a off -i ${interface}
+    done
```
Comment (yosefe) [inline buildlib/az-network-corrupter.sh:32]: i think shutting off one random NIC should be enough

File: buildlib/pr/io_demo/az-stage-io-demo.yaml [hunk 6]
```diff
@@ -43,28 +48,36 @@ steps:
...
+    for interface in ${{ parameters.roce_iface }} ; do
+      net_device=$(ibdev2netdev | sed -ne 's/\(\w*\) port \([0-9]\) ==> '${interface}' .*/\1:\2/p')
+      net_devices="${net_devices} ${net_device}"
+    done
+    export UCX_NET_DEVICES=$(echo ${net_devices##*( )} | tr " ", ",")
     export UCX_TLS=${{ parameters.iodemo_tls }}
     export UCX_RNDV_THRESH=4k
     export LD_LIBRARY_PATH=$(workspace)/install/lib:$LD_LIBRARY_PATH
+    export UCX_LOG_LEVEL=info
+    count_of_task=$(echo ${{ parameters.roce_iface }} | wc -w)
```
Comment (yosefe) [inline buildlib/pr/io_demo/az-stage-io-demo.yaml:69]: count_of_task --> num_tasks

File: buildlib/pr/io_demo/az-stage-io-demo.yaml [hunk 7]
```diff
@@ -83,11 +96,22 @@ steps:
...
     analyzer_args="$analyzer_args -t 3"
+    if [ '${{ parameters.interference }}' == 'No' ] ; then
+      analyzer_args="$analyzer_args --no-allow-list"
+    fi
     analyzer_args="$analyzer_args ${{ parameters.analyzer_allow_list_args }}"
     python ${analyzer} ${analyzer_args}
   displayName: Analyze for ${{ parameters.name }}
   timeoutInMinutes: 2
 
+- task: PublishBuildArtifacts@1
```
Comment (yosefe) [inline buildlib/pr/io_demo/az-stage-io-demo.yaml:107]: nice

File: buildlib/pr/io_demo/io-demo.yml [hunk 8]
```diff
@@ -54,23 +54,31 @@ parameters:
...
       "active messages on CX6/DC with user memh":
         args: "-A -z"
-        initial_delay: 9999 # No interference
+        interference: 'No'
         duration: 120
         interface: $(roce_iface_cx6)
         tls: "dc_x"
         test_name: am_umemh_cx6_dc
         extra_run_args: ""
+      "Multy HCAs: tag match on CX4 and CX6 RC ":
```
Comment (yosefe) [inline buildlib/pr/io_demo/io-demo.yml:71]: Multy -> Multi

File: buildlib/pr/io_demo/io-demo.yml [hunk 9]
```diff
@@ -17,50 +17,65 @@ parameters:
...
-      "active messages on CX6/DC with data validation":
-        args: "-A -q"
-        duration: 600
-        interface: $(roce_iface_cx6)
-        tls: "dc_x"
-        test_name: am_validate_cx6_dc
-      "tag match on CX4 with user memh":
-        args: "-z"
         initial_delay: 9999 # No interference
         duration: 120
```
[the remark this answers is not in the document]
Comment (yosefe) [reply buildlib/pr/io_demo/io-demo.yml:74]: 60 ?

File: test/apps/iodemo/run_io_demo.sh [hunk 10]
```diff
@@ -132,7 +132,7 @@ usage()
 	echo "  -h|--help                   Show this help message"
 	echo "  -v|--verbose                Turn on verbosity"
 	echo "  -H|--hostlist <h1>,<h2>,..  List of host names to run on"$(show_default_value host_list)
-	echo "  -i|--netif <n>              Network interface to use"$(show_default_value net_if)
+	echo "  -i|--netif <n>              Network interfaces to use"$(show_default_value net_ifs)
```
Comment (yosefe) [inline test/apps/iodemo/run_io_demo.sh:135]: `echo "  -i|--netif <n1,n2>              Comma-separated list of network interfaces to use"$(show_default_value net_ifs)`

File: test/apps/iodemo/run_io_demo.sh [hunk 11]
```diff
@@ -454,10 +457,13 @@ make_scripts()
 	client_connect_list=""
 	for host in $(split_list ${host_list})
 	do
-		for ((i=0;i<${num_servers_per_host[${host}]};++i))
+		for net_if in $(split_list ${net_ifs})
 		do
-			port_num=$((base_port_num + i))
-			client_connect_list+=" ${ip_address_per_host[${host}]}:${port_num}"
+			for ((i=0;i<${num_servers_per_host[${host}]};++i))
+			do
```
Comment (yosefe) [inline test/apps/iodemo/run_io_demo.sh:463]: better swap the for loop in line 462 and the one in line 460

File: buildlib/az-network-corrupter.sh (hunk 5 above)
Comment (avildema) [reply buildlib/az-network-corrupter.sh:32]: here I do it
--one - choose one random interface
-a off - turn off

File: buildlib/pr/io_demo/az-stage-io-demo.yaml [hunk 12]
```diff
@@ -13,13 +13,18 @@ parameters:
   default: 20
 - name: analyzer_allow_list_args
   default: ''
+- name: interference
+  default: 'Yes'
```
[the remark this answers is not in the document]
Comment (avildema) [reply buildlib/pr/io_demo/az-stage-io-demo.yaml:17]: we use constriction `echo "##vso[task.setvariable variable=interference]${{ parameters.interference }}"` and it is not support Boolean

**Review by yosefe — DISMISSED** [review]
pls squash

**Review by yosefe — APPROVED** [review]

**Pull request merged**